### DISCUSSION TITLE
v1.4.0

### DISCUSSION THREAD

**Discussion by duncanmacmichael** [microsoft/WindowsAppSDK#3827] [Announcements]
## Windows App SDK 1.4.0 (1.4.230822000)

WinAppSDK 1.4.0 is the latest stable version of WinAppSDK providing new features and improvements to the platform.

**New features include:**
- XAML Islands are no longer experimental.
- A new list control called `ItemsView`.
- Ease-of-use updates to `DesktopAcrylicController` and `MicaController`.

**Additional updates include:**
- Numerous performance, security, accessibility and reliability fixes to improve the quality of your app.
- Updated WinUI Gallery, Template Studio, and VS WinUI templates.

**To see everything that's new and changed, see the full [Windows App SDK 1.4 release notes](https://aka.ms/windowsappsdk/1.4/1.4.0/release-notes).**  

**See the [WinAppSDK 1.4 milestone](https://github.com/microsoft/WindowsAppSDK/milestone/15?closed=1) and [WinUI 3 in WinAppSDK 1.4 milestone](https://github.com/microsoft/microsoft-ui-xaml/milestone/19?closed=1) for a list of issues addressed in this release.**

### Try it out
- Download the [1.4.230822000 NuGet](https://www.nuget.org/packages/Microsoft.WindowsAppSDK/1.4.230822000) package to use WinAppSDK 1.4 in your app.
- Download and update the [WinUI Gallery](https://www.microsoft.com/store/productId/9P3JFPWWDZRC) to see the WinUI 3 updates firsthand.

### Getting Started 

To get started using Windows App SDK to develop Windows apps, check out the following documentation:    
- [Install developer tools ](https://docs.microsoft.com/windows/apps/windows-app-sdk/set-up-your-development-environment).     
- [Create a WinUI 3 app](https://docs.microsoft.com/windows/apps/winui/winui3/create-your-first-winui3-app).   
- [Continue your development journey](https://docs.microsoft.com/windows/apps/develop).   

<hr /><em>This discussion was created from the release <a href='https://github.com/microsoft/WindowsAppSDK/releases/tag/v1.4.0'>v1.4.0</a>.</em>

**Comment by kmgallahan** [comment]
> Updated WinUI Gallery

ETA on that? Looking forward to seeing the new `ItemsView` in action (if it was added).

**Reply by Mysterious-Dev** [reply]
Neither WinUi Gallery nor Template Studio have released "general public" versions yet. I think I saw that WinUI Gallery is already under version 1.4 of WASDK (a commit on this was pushed), but Template Studio hasn't had any changes yet.

**Reply by duncanmacmichael** [reply]
The WinUI Gallery is going through the release process now, sorry that it lagged behind a bit! Should be published very soon.

**Comment by Diegorro98** [comment]
What is needed to get the Customize button for Widgets? I already implemented `IWidgetProvider2` right next to `IWidgetProvider` on my implementation of `WidgetProvider`, but nothing seems to change, maybe Windows Web Experience Pack needs to be updated? Or maybe a Windows Insider build is required?

**Comment by beervoley** [comment]
> What is needed to get the Customize button for Widgets? I already implemented `IWidgetProvider2` right next to `IWidgetProvider` on my implementation of `WidgetProvider`, but nothing seems to change, maybe Windows Web Experience Pack needs to be updated? Or maybe a Windows Insider build is required?

You need to declare your Widget customizable in the appxmanifest declaration:)
```
<WidgetProvider>
   <Definition IsCustomizable="true"> ... </Definition>
</WidgetProvider>
 ```
Docs for this are coming :)

**Reply by Diegorro98** [reply]
Any update on docs?

**Reply by drewbatgit** [reply]
The docs for this feature are currently in draft and should be available publicly sometime next week.

**Reply by beervoley** [reply]
@Diegorro98 docs are up: https://learn.microsoft.com/en-us/windows/apps/develop/widgets/implement-widget-provider-win32#implementing-widget-customization

**Comment by james1301** [comment]
Hi now that XAML Islands is not experimental anymore. How do we use it it in WPF? I have had a look around and I'm pretty confused, we have what seems like some ancient XAML Islands documentation [here](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands).

Digging into the release notes it says that there is no convenient wrapper for WPF or WinForms. So can you do it, how? If this is XAML Islands for C++ is no longer experimental please do identify it as that. When is this going to be fully released for all platforms?

**Reply by DarranRowe** [reply]
This would be why the release notes mentioned that there are no convenience wrappers. Until those appear, you have to use Xaml Islands like you are hosting a Windows API control, from what I can see.
The big thing to remember is that the WinUI 3 Xaml Islands uses a WindowId to identify the window. When you create the DesktopWindowXamlSource instance, you then call [Initialize](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.hosting.desktopwindowxamlsource.initialize?view=windows-app-sdk-1.4#microsoft-ui-xaml-hosting-desktopwindowxamlsource-initialize(microsoft-ui-windowid)) to provide the parent to the Xaml content. You can get the underlying content using the [SiteBridge](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.hosting.desktopwindowxamlsource.sitebridge?view=windows-app-sdk-1.4) property. This gives you a [DesktopChildSiteBridge](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.content.desktopchildsitebridge?view=windows-app-sdk-1.4) instance, the WindowId you get from this is the identifier for the Xaml Island itself. You can convert between a HWND and WindowId by using GetWindowFromWindowId and GetWindowIdFromWindow which can be found in Microsoft.UI.Win32Interop.
Basically, this means that you should look into doing it the C++ way and trying to fit it in to WPF for the time being. It shouldn't take too long for the more adventurous people to write their own wrappers. The C++ method of using WinUI 3 Xaml Islands is pretty similar to the Windows 10 Xaml Islands implementation.

**Reply by castorix** [reply]
> How do we use it it in WPF?

With [DesktopWindowXamlSource ](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.hosting.desktopwindowxamlsource?view=windows-app-sdk-1.4)
For a minimal test, you set the _Content_ and call _Initialize_  (from _Microsoft.UI.Win32Interop.GetWindowIdFromWindow_) to set it as child of the main window

**Reply by james1301** [reply]
Thanks @DarranRowe and @castorix this sounds not ready for me then really in terms of WPF and WinForms. I just want to review whether we could make use of it or not really in our current applications and it's current form that would just be a no. What am I to follow to know when this is ready for WPF?

**Comment by castorix** [comment]
> Thanks @DarranRowe and @castorix this sounds not ready for me then really in terms of WPF and WinForms. I just want to review whether we could make use of it or not really in our current applications and it's current form that would just be a no. What am I to follow to know when this is ready for WPF?

WinForms/C# seems to work, like in the link of a test sample ([CSharp_WinUI3_DesktopWindowXamlSource](https://github.com/castorix/CSharp_WinUI3_DesktopWindowXamlSource)) I posted in another thread (https://github.com/microsoft/microsoft-ui-xaml/issues/7029)
But I still have problems with WPF, by using nearly the same code (I tested in _OnSourceInitialized_ or _OnContentRendered_ instead of _Form1_Load_...)

**Comment by castorix** [comment]
> But I still have problems with WPF, by using nearly the same code (I tested in _OnSourceInitialized_ or _OnContentRendered_ instead of _Form1_Load_...)

I just found the problem, thanks to https://github.com/microsoft/microsoft-ui-xaml/issues/8806 ,
 it was "Enable in-app Toolbar" ...

**Reply by JochemPalmsens** [reply]
@duncanmacmichael The "in-app toolbar" is heavily used by our front-end developers. So having to disable it to prevent a crash at the start of the app (see linked issue in the message above) is blocking us from upgrading to WinAppSDK1.4. Do you know when this will be fixed?

**Reply by duncanmacmichael** [reply]
Honestly I don't, sorry - I'll have to ask the engineering folks who work on the bugs. @bpulliam do we have any updates on this bug?

**Reply by duncanmacmichael** [reply]
Update: We're looking into this and will update you as soon as we have new info, thanks for your patience.

**Comment by duncanmacmichael** [comment]
Hi folks, wanted to let you know that we had a late addition to the release notes for v1.4.0 about new custom titlebar features! You can check out the notes here:

https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#custom-titlebar--appwindow-titlebar-merger

And check out the code and screenshot here:

https://github.com/microsoft/WinUI-Gallery/pull/1360

You'll be able to see these changes in the WinUI Gallery app on the next release, or alternatively you can see them now by building the repo. Take it for a spin and let us know what you think!
